tsuba: keep all type information around

tylershunt opened the pull request:
As others have noted (@witchel) it is desirable to have type information
for properties that are not loaded. This PR adds that facility along
with some mechanism for the sane management of same.

I elected not to modify our metadata format. Although the metadata
files are surely a logical place for this, I'm loath to change the
format so often. Instead, this code reads the type of columns out of
the parquet files for unloaded properties. This is slower but requires
no metadata changes.

We could always add it to the metadata later as an optimization if this
way of doing things is deemed too slow.

Signed-off-by: Tyler Hunt <thunt@katanagraph.com>

insertinterestingnamehere commented on conda_recipe/conda_build_config.yaml:
```diff
@@ -0,0 +1,12 @@
+pin_run_as_build:
+  boost-cpp: x.x
+  arrow-cpp: x.x
+  eigen: x.x
```
Eigen may actually be just a build-time dependency since it's a header-only library. In fact, I think it's currently only used in the C++ version of matrix completion, so we may not need it at all for the conda package.

insertinterestingnamehere commented on conda_recipe/conda_build_config.yaml:
```diff
@@ -0,0 +1,12 @@
+  boost-cpp: x.x
+  arrow-cpp: x.x
+  eigen: x.x
+  fmt: x.x
```
libfmt's api is probably stable enough at this point that we could just pin the major version. Same goes for howardhinnant_date.

insertinterestingnamehere commented on conda_recipe/meta.yaml:
```diff
@@ -8,41 +8,39 @@ source:
-    - aws-sdk-cpp>=1.7.164
-    - eigen>=3.3.7
-    - fmt>=6.2.1,<7
+    - boost-cpp>=1.71,<2.0a0
```
Cool trick. I didn't know adding the alpha on was needed, but it makes sense.

Commit:
tsuba: keep all type information around

As others have noted (@witchel) it is desirable to have type information
for properties that are not loaded. This PR adds that facility along
with some mechanism for the sane management of same.

I elected not to modify our metadata format. Although the metadata
files are surely a logical place for this, I'm loath to change the
format so often. Instead, this code reads the type of columns out of
the parquet files for unloaded properties. This is slower but requires
no metadata changes.

We could always add it to the metadata later as an optimization if this
way of doing things is deemed too slow.

Signed-off-by: Tyler Hunt <thunt@katanagraph.com>

witchel commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -465,14 +465,26 @@ class KATANA_EXPORT PropertyGraph {
   std::string ReportDiff(const PropertyGraph* other) const;
 
+  /// get the schema for loaded node properties
   std::shared_ptr<arrow::Schema> node_schema() const {
```
I see the backward compatibility advantage of `node_schema` and `full_node_schema`, but I feel like moving forward, `loaded_node_schema` and `full_node_schema` would be more logical.  But maybe that is just how I vibe.

tylershunt replied:
Yeah that seems like a good change.

witchel commented on libtsuba/src/ParquetReader.cpp:
```diff
@@ -209,6 +222,14 @@ class BlockedParquetReader {
+  Result<std::shared_ptr<arrow::Schema>> ReadSchema() {
+    KATANA_CHECKED(EnsureReader(0));
+    std::shared_ptr<arrow::Schema> schema;
+    KATANA_CHECKED(parquet::arrow::FromParquetSchema(
+        readers_[0]->parquet_reader()->metadata()->schema(), &schema));
```
Is there a single parquet file with the full schema?  I thought you were going to get the type from each parquet file (property column) and combine them.

tylershunt replied:
There isn't, but this is a more general Reader class that just gets the schema from whatever file you pass it. So I do call this once per property file to get the type.

witchel commented on libtsuba/src/RDGCore.cpp:
```diff
@@ -91,22 +117,46 @@ namespace tsuba {
+}
+
+katana::Result<void>
+RDGCore::EnsureNodeTypesLoaded(const katana::Uri& rdg_dir) {
```
So these load all properties.

tylershunt replied:
Just their types.

witchel commented on libtsuba/src/RDGPartHeader.h:
```diff
@@ -87,6 +92,16 @@ class PropStorageInfo {
   const std::string& path() const { return path_; }
+  const std::shared_ptr<arrow::DataType>& type() const { return type_; }
+
+  // since we don't have type info in the header don't know the
+  // type when this would have been constructed. Allow others to
+  // fix up the type in this case, required until we can get the type
+  // from PartHeader files
+  void set_type(std::shared_ptr<arrow::DataType> type) {
```
Ah, this is why you are passing around the type to `WasModified`.  Yeah, this cost is probably worth the price of not changing the on-disk format.  Though I fear changing the on-disk format will only get more difficult...

witchel reviewed (approved):
I would say this PR is the stuff of dreams, but my dreams aren't this detailed.

tylershunt commented:
Requires the sister PR https://github.com/KatanaGraph/katana-enterprise/pull/1650 to fix enterprise after this merges.
